src/shared/cryptography.c

```diff
@@ -0,0 +1,151 @@
+        return ERROR_INVALID_DATA;
+    default:
+        last_error = GetLastError();
+        merror("WinVerifyTrust returned %lX GetLastError returned %lX", status, last_error);
+        return ERROR_INVALID_DATA;
+    }
+}
+
+DWORD get_file_hash(const wchar_t *path, BYTE **hash, DWORD *hash_size)
+{
+    DWORD result = ERROR_SUCCESS;
+
+    // Get full path if path is a relative path.
+    // This is needed because CryptCATAdminCalcHashFromFileHandle only accepts full paths.
+    // If path is already a full path, GetFullPathName will return the same path.
+    wchar_t full_path[MAX_PATH];
+
+    if (GetFullPathNameW(path, MAX_PATH, full_path, NULL)) {
+        // Open file for hash calculation.
+        HANDLE handle_file = CreateFileW(full_path, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+        if (handle_file != INVALID_HANDLE_VALUE) {
+
+            // Calculate hash of the file.
+            if (CryptCATAdminCalcHashFromFileHandle(handle_file, hash_size, NULL, 0)) {
+
+                if (*hash_size == 0) {
+                    result = ERROR_INVALID_DATA;
+                    merror("CryptCATAdminCalcHashFromFileHandle failed with error %lu", GetLastError());
+                } else {
+                    os_calloc(1, *hash_size, *hash);
+
+                    if (!CryptCATAdminCalcHashFromFileHandle(handle_file, hash_size, *hash, 0)) {
+                        result = GetLastError();
+                        merror("CryptCATAdminCalcHashFromFileHandle failed with error %lu", result);
+                        os_free(*hash);
+                    }
+                }
+            }
+        } else {
+            result = ERROR_FILE_NOT_FOUND;
+            mdebug2("CreateFileW failed with error %lu path %S", GetLastError(), path);
+        }
+        // Close file handle.
+        CloseHandle(handle_file);
+
+    } else {
+        result = ERROR_INVALID_DATA;
+        merror("GetFullPathNameW failed with error %lu", GetLastError());
+    }
+
+    return result;
+}
+
+DWORD verify_hash_catalog(wchar_t *file_path)
+{
+    HCATADMIN catalog_administrator = NULL;
+    HCATINFO catalog_context = NULL;
+    GUID policy_GUID = DRIVER_ACTION_VERIFY;
+    BYTE *hash = NULL;
+    DWORD hash_size = 0;
+    DWORD result;
+
+    if (result = get_file_hash(file_path, &hash, &hash_size), ERROR_SUCCESS == result) {
+        if (CryptCATAdminAcquireContext(&catalog_administrator, &policy_GUID, 0)) {
+            // Search for the catalog file.
+            // If the file is not signed, the function returns NULL.
+            // If the file is signed, the function returns a handle to the catalog file.
+
+            if (catalog_context = CryptCATAdminEnumCatalogFromHash(catalog_administrator, hash, hash_size, 0, NULL), catalog_context) {
+                CryptCATAdminReleaseCatalogContext(catalog_administrator, catalog_context, 0);
+            } else {
+                result = GetLastError();
```

Comment:
I can't find information about `CryptCATAdminEnumCatalogFromHash`.  But here we are relying %100 that in case of failure, it'll set the error variable. If it doesn't, we could return a success code incorrectly. 

I'd write this else like this 
```c++
} else {
                result = ERROR_INVALID_DATA;
                merror("CryptCATAdminEnumCatalogFromHash failed with error %lu",  GetLastError());
            }


```

Reply:
good catch

Reply:
![Uploading image.png…]()